src/diagnostics/gnu-ld.ts

```diff
@@ -4,33 +4,167 @@
 
 import * as vscode from 'vscode';
 
-import { FeedLineResult, oneLess, RawDiagnosticParser } from './util';
+import { oneLess, RawDiagnostic, RawDiagnosticParser, RawRelated, FeedLineResult } from './util';
 
-export const REGEX = /^(.*):(\d+)\s?:\s+(.*[^\]])$/;
+enum MatchType {
+    Full,
+    File,
+    Line,
+    Column,
+    Severity,
+    Message
+}
+
+const regexPatterns: RegexPattern[] = [
+    {   // path/to/ld[.exe]:[ ]path/to/file:line: severity: message
+        regexPattern: /^(?:.*ld(?:\.exe)?:)(?:\s*)?(.+):(\d+):\s+(?:fatal )?(\w+):\s+(.+)/,
+        matchTypes: [MatchType.Full, MatchType.File, MatchType.Line, MatchType.Severity, MatchType.Message]
+    },
+    {   // path/to/ld[.exe]:[ ]path/to/file:line: message
+        regexPattern: /^(?:.*ld(?:\.exe)?\:)(?:\s*)?(.+):(\d+):\s+(.+)/,
+        matchTypes: [MatchType.Full, MatchType.File, MatchType.Line, MatchType.Message]
+    },
+    {   // path/to/ld[.exe]: severity: message
+        regexPattern: /^(.*ld(?:\.exe)?):\s+(?:fatal )?(\w+):\s+(.+)/,
+        matchTypes: [MatchType.Full, MatchType.File, MatchType.Severity, MatchType.Message]
+    },
+    {   // path/to/ld[.exe]: message (without trailing colon)
+        regexPattern: /^(.*ld(?:\.exe)?):\s+(.+)(?<!:)$/,
+        matchTypes: [MatchType.Full, MatchType.File, MatchType.Message]
+    },
+    {   // /path/to/file:line: message (without "[fatal] severity:" or trailing colon)
+        regexPattern: /^(.+?):(\d+):\s+(?!fatal\s+\w+:)(?!\w+:)(.+)(?<!:)$/,
+        matchTypes: [MatchType.Full, MatchType.File, MatchType.Line, MatchType.Message]
+    }
+];
+
+interface RegexPattern {
```

Comment:
@0xemgy Could we possibly share the `RegexPattern` interface and the `MatchType` interface rather than doubly defining it?

Reply:
Moved both interface to the util.ts and imported them.